code/espurna/sensors/ADE7953Sensor.h

```diff
@@ -0,0 +1,215 @@
+                (float)active_power2 / (ADE7953_PREF / 10)
+            );
+        }
+
+        void writeFloat(unsigned int relay, float current, float power) {
+            auto& reading_ref = _readings.at(relay); 
+            reading_ref.current = current;
+            reading_ref.power = power;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {        
+            if (index == 0) return _voltage;            
+            int relay = (index - 1) / ADE7953_TOTAL_DEVICES;	
+            index = index % ADE7953_TOTAL_DEVICES;
+            if (index == 0) return _readings[relay].energy;
+            if (index == 1) return _readings[relay].current;
+            if (index == 2) return _readings[relay].power;     
+            return 0;
+        }
+
+    protected:  
+        void _init() {                     
+            delay(100);                     // Need 100mS to init ADE7953
+            Ade7953Write(0x102, 0x0004);    // Locking the communication interface (Clear bit COMM_LOCK), Enable HPF
+            Ade7953Write(0x0FE, 0x00AD);    // Unlock register 0x120
+            Ade7953Write(0x120, 0x0030);    // Configure optimum setting       
+                    
+            _ready = true;
+        }
+
+        int Ade7953RegSize(uint16_t reg) {
+            int size = 0;
+            switch ((reg >> 8) & 0x0F) {
+                case 0x03:
+                size++;
+                case 0x02:
+                size++;
+                case 0x01:
+                size++;
+                case 0x00:
+                case 0x07:
+                case 0x08:
+                size++;
+            }
+            return size;
+        }
+            
+        void Ade7953Write(uint16_t reg, uint32_t val) {
+            int size = Ade7953RegSize(reg);
+            if (size) {
+                Wire.beginTransmission(_address);
+                Wire.write((reg >> 8) & 0xFF);
+                Wire.write(reg & 0xFF);
+                while (size--) {
+                Wire.write((val >> (8 * size)) & 0xFF);  // Write data, MSB first
+                }
+                Wire.endTransmission();
+                delayMicroseconds(5);    // Bus-free time minimum 4.7us
+            }
+        }
+            
+        uint32_t Ade7953Read(uint16_t reg) {
+            uint32_t response = 0;
+            
+            int size = Ade7953RegSize(reg);
+            if (size) {
+                Wire.beginTransmission(_address);
+                Wire.write((reg >> 8) & 0xFF);
+                Wire.write(reg & 0xFF);
+                Wire.endTransmission(0);
+                Wire.requestFrom(_address, size);
+                if (size <= Wire.available()) {
+                for (int i = 0; i < size; i++) {
+                    response = response << 8 | Wire.read();   // receive DATA (MSB first)
+                }
+                }
+            }
+            return response;
+        }
+
+    typedef struct {    
```

Comment:
*just a style note*
because it is c++, this can be just `struct reading_t {}`

Reply:
Thanks! I'm not a c/c++ writer!

Reply:
doesn't compile for me! example: `'struct ADE7953Sensor::reading_t' has no member named 'current'`